material-maker: init at 1.3

aurreland opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Material Maker is a Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models.

https://www.materialmaker.org/

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

aurreland committed:
material-maker: init at 1.3

Co-Authored-By: d-brasher <175485311+d-brasher@users.noreply.github.com>
Co-Authored-By: Fedi Jamoussi <Fedi.Jamoussi@protonmail.ch>

aurreland committed:
maintainers: add aurreland

maintainer: add aurreland

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/package-godot-projects-to-nixpkgs/49923/15

aurreland commented:
Its running checks for darwin, I've not package it for darwin because i doesnt have a mac and it would be pointless as they could install it via a .dmg. Should i try to add darwin compatibity or maybe enable something so it doesnt bother about it ?

eljamm commented:
If it can be done, I don't see why we shouldn't package this for mac as well. We should refactor the current derivation first, though, then we'll look into doing that.

I'm currently in the process if reviewing this, so you can expect my suggestions shortly.

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,140 @@
+    udev
+  ];
+
+  meta = with lib; {
+    homepage = "https://www.materialmaker.org/";
+    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models.";
+    license = licenses.mit;
+    platforms = [ "x86_64-linux" ];
+    maintainers = with maintainers; [ aurreland ];
```
```suggestion
  meta = {
    homepage = "https://www.materialmaker.org/";
    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models.";
    license = lib.licenses.mit;
    platforms = lib.platforms.x86_64-linux;
    maintainers = with lib.maintainers; [ aurreland ];
```

See https://github.com/NixOS/nixpkgs/issues/208242

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,140 @@
+
+  meta = with lib; {
+    homepage = "https://www.materialmaker.org/";
+    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models.";
```
```suggestion
    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models";
```

As specified in https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,140 @@
+  src = if stdenv.isLinux then fetchurl {
+    url = "https://github.com/RodZill4/material-maker/releases/download/1.3/material_maker_1_3_linux.tar.gz";
+    hash = "sha256-Y8+ZwXy3zqnsxqqaZeVgFSxLzmUkq+rBzbq8tEDc8/g=";
+  } else throw "unsupported platform";
```
Use `fetchzip` instead of `fetchurl` since it auto-unpacks. We can also reuse the defined attributes like `pname`, `version` and `versionUnderscore`

```suggestion
  pname = "material-maker";
  inherit version;

  src =
    if stdenv.isLinux then
      fetchzip {
        url = "https://github.com/RodZill4/${pname}/releases/download/${version}/material_maker_${versionUnderscore}_linux.tar.gz";
        hash = "sha256-WEu5gVfnswB5zYzu3leOL+hKOBzJbn48gHQKshlfOh4=";
      }
    else
      throw "unsupported platform";
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,144 @@
+  udev,
+}:
+stdenv.mkDerivation rec {
+  pname = "material-maker";
+  version = "1.3";
+
+  src =
+    if stdenv.isLinux then
+      fetchurl {
+        url = "https://github.com/RodZill4/material-maker/releases/download/1.3/material_maker_1_3_linux.tar.gz";
+        hash = "sha256-Y8+ZwXy3zqnsxqqaZeVgFSxLzmUkq+rBzbq8tEDc8/g=";
+      }
+    else
+      throw "unsupported platform";
```
Use `fetchzip` instead of `fetchurl` since it auto-unpacks it. Also, re-use the attributes like `pname`, `version` and `versionUnderscore` for the url

```suggestion
  pname = "material-maker";
  inherit version;

  src =
    if stdenv.isLinux then
      fetchzip {
        url = "https://github.com/RodZill4/${pname}/releases/download/${version}/material_maker_${versionUnderscore}_linux.tar.gz";
        hash = "sha256-WEu5gVfnswB5zYzu3leOL+hKOBzJbn48gHQKshlfOh4=";
      }
    else
      throw "unsupported platform";
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,144 @@
+    else
+      throw "unsupported platform";
+
+  icon = fetchurl {
+    url = "https://github.com/RodZill4/material-maker/raw/1.3/icon.png";
+    hash = "sha256-cu4aD8WbeYuISeub0Mt7Z/2e4/9BZF+QwsqYFo82Ge8=";
+  };
```
We don't have to get the icon as it's already present in the package, just not in an obvious place

```suggestion
```

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,140 @@
+  zlib,
+  udev,
+}:
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

See https://github.com/NixOS/nixpkgs/issues/315337#issuecomment-2138814370

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,144 @@
+    godot3-headless
+  ];
+
+  buildInputs = [
+    alsa-lib
+    libGL
+    libGLU
+    libX11
+    libXcursor
+    libXext
+    libXfixes
+    libXi
+    libXinerama
+    libXrandr
+    libXrender
+    zlib
+    udev
+  ];
```
I think only these are the necessary runtime packages as the `autoPatchelfHook` doesn't seem to need anything else.

```suggestion
  buildInputs = [
    libGL
    libGLU
    libX11
    libXcursor
    libXext
    libXfixes
    libXi
    libXinerama
    libXrandr
    libXrender
  ];
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,144 @@
+    udev
+  ];
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "material-maker";
+      exec = "material-maker";
+      icon = "material-maker";
+      desktopName = "Material Maker";
+      comment = meta.description;
+      categories = [
+        "Graphics"
+        "Development"
+      ];
+      keywords = [ "material" ];
+    })
+  ];
```
We can use some information from the upstream [desktop file](https://github.com/RodZill4/material-maker/blob/0268009d02061c566c281a0d65adde67f2f1caa1/material_maker/misc/linux/io.github.RodZill4.Material-Maker.desktop).

Also, we don't have to call `makeDesktopItem` from a list as we only have one desktop file.

```suggestion
  desktopItems = makeDesktopItem {
    name = "material-maker";
    exec = "material-maker";
    icon = "material-maker";
    genericName = "Procedural texture generation and 3D model painting tool";
    desktopName = "Material Maker";
    comment = meta.description;
    terminal = false;
    categories = [
      "Graphics"
      "3DGraphics"
    ];
    keywords = [ "material" ];
  };
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,144 @@
+    })
+  ];
+
+  unpackPhase = ''
+    mkdir -p src
+    tar -xzvf $src -C src
+  '';
+
```
We don't need this anymore since we can use `fetchzip`, which auto-unpacks the file.

```suggestion
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,144 @@
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    install -D -m 755 ./src/material_maker_1_3_linux/material_maker.x86_64 $out/libexec/material-maker
+    install -D -m 644 ./src/material_maker_1_3_linux/material_maker.pck $out/libexec/material-maker.pck
+
+    mkdir -p $out/share/icons/hicolor/256x256/apps/
+    install -D -m 644 $icon -t $out/share/icons/hicolor/256x256/apps/
+
+    cp -r ./src/material_maker_1_3_linux/doc $out/libexec
+    chmod -R 755 $out/libexec/doc
+    cp -r ./src/material_maker_1_3_linux/environments $out/libexec
+    chmod -R 755 $out/libexec/environments
+    cp -r ./src/material_maker_1_3_linux/examples $out/libexec
+    chmod -R 755 $out/libexec/examples
+    cp -r ./src/material_maker_1_3_linux/export $out/libexec
+    chmod -R 755 $out/libexec/export
+    cp -r ./src/material_maker_1_3_linux/library $out/libexec
+    chmod -R 755 $out/libexec/library
+    cp -r ./src/material_maker_1_3_linux/meshes $out/libexec
+    chmod -R 755 $out/libexec/meshes
+    cp -r ./src/material_maker_1_3_linux/nodes $out/libexec
+    chmod -R 755 $out/libexec/nodes
+
+    install -d -m 755 $out/bin
+    ln -s $out/libexec/material-maker $out/bin/material-maker
+
+    runHook postInstall
+  '';
```
We can simplify copying the required directories using a for loop. I don't think it's really necessary to modify their permissions, though. Feel free to also set those if you think otherwise, though.

Also, the icon exists in the source under `$src/doc/_static/icon.png` which we can just copy that over.

```suggestion
  installPhase = ''
    runHook preInstall

    install -D -m 755 $src/material_maker.x86_64 $out/libexec/material-maker
    install -D -m 644 $src/material_maker.pck $out/libexec/material-maker.pck

    for dir in examples library nodes doc environments meshes export; do
      cp -R $src/$dir $out/libexec
    done

    install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/

    mkdir -p $out/bin
    ln -s $out/libexec/material-maker $out/bin/material-maker

    runHook postInstall
  '';
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,144 @@
+    runHook postInstall
+  '';
+
+  runtimeDependencies = map lib.getLib [
+    alsa-lib
+    libpulseaudio
+    udev
+  ];
+
```
Don't think we really need this.

```suggestion
```

eljamm commented:
As a tip, you can batch commit multiple suggestions from the `Files Changed` tab, so you don't have to do it for each of them.

See [Incorporating feedback in your pull request - GitHub Docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request)

aurreland commented:
Thanks for the huge work eljamm and d-brasher, i will look into how we could add support for darwin.

eljamm commented:
Since we're using `finalAttrs`, you'll have to precede every variable name that was defined inside the derivation with that with that.

So:
- ${version} -> ${version} (stays version because we defined it in the `let in` block
- ${pname} -> ${finalAttrs.pname}
- ${meta.description} -> ${finalAttrs.meta.description}
- ...

aurreland commented:
Good to know will change to this

eljamm commented:
Just as a plus, since we're inheriting `version` in the derivation, we can also do `${finalAttrs.version}`, but that won't work for `versionUnderscore` since it wasn't inherited.

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,117 @@
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    homepage = "https://www.materialmaker.org/";
+    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models";
+    license = lib.licenses.mit;
+    platforms = lib.platforms.x86_64-linux;
+    maintainers = with libmaintainers; [ aurreland ];
+    mainProgram = "material-maker";
+  };
```
Somehow, `with lib;` is still there, and `lib.maintainers` is wrong

```suggestion
  meta = {
    homepage = "https://www.materialmaker.org/";
    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models";
    license = lib.licenses.mit;
    platforms = lib.platforms.x86_64-linux;
    maintainers = with lib.maintainers; [ aurreland ];
    mainProgram = "material-maker";
  };
```

aurreland replied:
My bad, didnt pay attention

eljamm replied:
No worries 😁

aurreland commented:
Should i pass it to ready now ?

eljamm commented:
Not yet, we still need to add darwin support.

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,117 @@
+    homepage = "https://www.materialmaker.org/";
+    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models";
+    license = lib.licenses.mit;
+    platforms = lib.platforms.x86_64-linux;
```
There is no `lib.platforms.x86_64-linux;`

```suggestion
    platforms = lib.intersectLists lib.platforms.linux lib.platforms.x86_64;
```

aurreland commented:
How could we add support for darwin, because we are using binary do they work out of the box ? or should we try to remake the package to be compiled like [lorien](https://github.com/NixOS/nixpkgs/blob/b2dbd6bf70529c30c64eb51c75607b2065d99df6/pkgs/applications/graphics/lorien/default.nix#L4) does ?

eljamm commented:
If the platform is Darwin, we just copy the source to the Applications folder, as far as I know, at least.

aurreland commented:
I will try tomorrow on an old mac i got, but i won't be able to do much till 24/08 as i am going on a vacation.

eljamm commented:
Well, in that case you can open this PR for review because I think this is good enough to be merged.

If we can figure out how to package and test this on mac, then we'll include it here tomorrow, else we can add that in a separate PR if this gets merged first.

eljamm reviewed (approved):
LGTM

d-brasher commented:
I think it could be a good thing to squash the commits in order to have just
`maintainers: add aurreland` (mind the `s`!) and `material-maker: init at 1.3`

See https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

eljamm commented:
I totally forgot about that :facepalm:

Thanks!

eljamm commented:
If I'm correct, this should be enough to get the build working for darwin:

<details>
<summary> support-darwin.patch </summary>

```diff
---
 pkgs/by-name/ma/material-maker/package.nix | 69 ++++++++++++++--------
 1 file changed, 43 insertions(+), 26 deletions(-)

diff --git a/pkgs/by-name/ma/material-maker/package.nix b/pkgs/by-name/ma/material-maker/package.nix
index ad2bb357a433..03faef297c85 100644
--- a/pkgs/by-name/ma/material-maker/package.nix
+++ b/pkgs/by-name/ma/material-maker/package.nix
@@ -23,19 +23,29 @@
   libXrender,
 }:
 let
+  pname = "material-maker";
   version = "1.3";
   versionUnderscore = "${lib.replaceStrings [ "." ] [ "_" ] version}";
+
+  sources = {
+    linux = fetchzip {
+      url = "https://github.com/RodZill4/${pname}/releases/download/${version}/material_maker_${versionUnderscore}_linux.tar.gz";
+      hash = "sha256-WEu5gVfnswB5zYzu3leOL+hKOBzJbn48gHQKshlfOh4=";
+    };
+    darwin = fetchzip {
+      url = "https://github.com/RodZill4/${pname}/releases/download/${version}/material_maker_${versionUnderscore}.dmg";
+      hash = "";
+    };
+  };
 in
 stdenv.mkDerivation (finalAttrs: {
-  pname = "material-maker";
-  inherit version;
+  inherit pname version;
 
   src =
-    if stdenv.isLinux then
-      fetchzip {
-        url = "https://github.com/RodZill4/${finalAttrs.pname}/releases/download/${finalAttrs.version}/material_maker_${versionUnderscore}_linux.tar.gz";
-        hash = "sha256-WEu5gVfnswB5zYzu3leOL+hKOBzJbn48gHQKshlfOh4=";
-      }
+    if stdenv.isLinux && stdenv.isx86_64 then
+      sources.linux
+    else if stdenv.isDarwin then
+      sources.darwin
     else
       throw "unsupported platform";
 
@@ -73,7 +83,7 @@ stdenv.mkDerivation (finalAttrs: {
     keywords = [ "material" ];
   };
 
-  buildPhase = ''
+  buildPhase = lib.optionalString stdenv.isLinux ''
     runHook preBuild
 
     # Cannot create file '/homeless-shelter/.config/godot/projects/...'
@@ -88,29 +98,36 @@ stdenv.mkDerivation (finalAttrs: {
     runHook postBuild
   '';
 
-  installPhase = ''
-    runHook preInstall
-
-    install -D -m 755 $src/material_maker.x86_64 $out/libexec/material-maker
-    install -D -m 644 $src/material_maker.pck $out/libexec/material-maker.pck
-
-    for dir in examples library nodes doc environments meshes export; do
-      cp -R $src/$dir $out/libexec
-    done
-
-    install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/
-
-    mkdir -p $out/bin
-    ln -s $out/libexec/material-maker $out/bin/material-maker
-
-    runHook postInstall
-  '';
+  installPhase =
+    ''
+      runHook preInstall
+    ''
+    + lib.optionalString stdenv.isLinux ''
+      install -D -m 755 $src/material_maker.x86_64 $out/libexec/material-maker
+      install -D -m 644 $src/material_maker.pck $out/libexec/material-maker.pck
+
+      for dir in examples library nodes doc environments meshes export; do
+        cp -R $src/$dir $out/libexec
+      done
+
+      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/
+
+      mkdir -p $out/bin
+      ln -s $out/libexec/material-maker $out/bin/material-maker
+    ''
+    + lib.optionalString stdenv.isDarwin ''
+      mkdir -p $out/Applications
+      cp -R *.app $out/Applications
+    ''
+    + ''
+      runHook postInstall
+    '';
 
   meta = {
     homepage = "https://www.materialmaker.org/";
     description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models";
     license = lib.licenses.mit;
-    platforms = lib.intersectLists lib.platforms.linux lib.platforms.x86_64;
+    platforms = lib.platforms.darwin ++ (lib.intersectLists lib.platforms.linux lib.platforms.x86_64);
     maintainers = with lib.maintainers; [ aurreland ];
     mainProgram = "material-maker";
   };
-- 
2.45.2
```

</details>

Just write this to a file (`support-darwin.patch`, for example), then `git apply support-darwin.patch`.

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,134 @@
+    keywords = [ "material" ];
+  };
+
+  buildPhase = lib.optionalString stdenv.isLinux ''
+    runHook preBuild
+
+    # Cannot create file '/homeless-shelter/.config/godot/projects/...'
+    export HOME=$TMPDIR
+
+    # Link the export-templates to the expected location. The --export commands
+    # expects the template-file at .../templates/3.2.3.stable/linux_x11_64_release
+    # with 3.2.3 being the version of godot.
+    mkdir -p $HOME/.local/share/godot
+    ln -s ${godot3-export-templates}/share/godot/templates $HOME/.local/share/godot
+
+    runHook postBuild
+  '';
+
```
We also don't need this as well. We're only linking the export templates, but we're not exporting anything.

```suggestion
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,134 @@
+  nativeBuildInputs = [
+    autoPatchelfHook
+    copyDesktopItems
+    godot3-headless
```
```suggestion
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,134 @@
+
+  # build deps
+  autoPatchelfHook,
+  godot3-export-templates,
+  godot3-headless,
```
I don't know how I missed this, but we don't need these at all 🙃

```suggestion
```

aurreland commented:
Should be squashed now ?

eljamm commented:
Yup

aurreland commented:
Should be now

eljamm commented:
Now it's good. Thanks :+1:

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,116 @@
+    };
+    darwin = fetchzip {
+      url = "https://github.com/RodZill4/${pname}/releases/download/${version}/material_maker_${versionUnderscore}.dmg";
+      hash = "";
```
Empty hash. Oops.

```suggestion
      hash = "sha256-6Jd1NBV5CwVjA/qHNfQ2szeHG7eFPX1vl+RXG5Lq6k4=";
```

aurreland commented:
We noticed at the same time lol, is it normal we aint got the same result for the hash ?

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,116 @@
+    libXfixes
+    libXi
+    libXinerama
+    libXrandr
+    libXrender
+  ];
+
+  desktopItems = makeDesktopItem {
+    name = "material-maker";
+    exec = "material-maker";
+    icon = "material-maker";
+    genericName = "Procedural texture generation and 3D model painting tool";
+    desktopName = "Material Maker";
+    comment = finalAttrs.meta.description;
+    terminal = false;
+    categories = [
+      "Graphics"
+      "3DGraphics"
+    ];
+    keywords = [ "material" ];
+  };
+
+  installPhase =
+    ''
+      runHook preInstall
+    ''
+    + lib.optionalString stdenv.isLinux ''
+      install -D -m 755 $src/material_maker.x86_64 $out/libexec/material-maker
+      install -D -m 644 $src/material_maker.pck $out/libexec/material-maker.pck
+
+      for dir in examples library nodes doc environments meshes export; do
+        cp -R $src/$dir $out/libexec
+      done
+
+      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/
+
+      mkdir -p $out/bin
+      ln -s $out/libexec/material-maker $out/bin/material-maker
+    ''
+    + lib.optionalString stdenv.isDarwin ''
```
```suggestion
  stdenvNoCC,
  fetchzip,

  # build deps
  autoPatchelfHook,
  copyDesktopItems,
  makeDesktopItem,

  # runtime deps
  libGL,
  libGLU,
  libX11,
  libXcursor,
  libXext,
  libXfixes,
  libXi,
  libXinerama,
  libXrandr,
  libXrender,
}:
let
  pname = "material-maker";
  version = "1.3";
  versionUnderscore = "${lib.replaceStrings [ "." ] [ "_" ] version}";

  sources = {
    linux = fetchzip {
      url = "https://github.com/RodZill4/${pname}/releases/download/${version}/material_maker_${versionUnderscore}_linux.tar.gz";
      hash = "sha256-WEu5gVfnswB5zYzu3leOL+hKOBzJbn48gHQKshlfOh4=";
    };
    darwin = fetchzip {
      url = "https://github.com/RodZill4/${pname}/releases/download/${version}/material_maker_${versionUnderscore}.dmg";
      hash = "07isbwcv42klh0y7wvn93hw4ms1girbxxvlcrmwh1cz7ay0vjjsq";
    };
  };
in
stdenvNoCC.mkDerivation (finalAttrs: {
  inherit pname version;

  src =
    if stdenvNoCC.isLinux && stdenvNoCC.isx86_64 then
      sources.linux
    else if stdenvNoCC.isDarwin then
      sources.darwin
    else
      throw "unsupported platform";

  nativeBuildInputs = [
    autoPatchelfHook
    copyDesktopItems
  ];

  buildInputs = [
    libGL
    libGLU
    libX11
    libXcursor
    libXext
    libXfixes
    libXi
    libXinerama
    libXrandr
    libXrender
  ];

  desktopItems = makeDesktopItem {
    name = "material-maker";
    exec = "material-maker";
    icon = "material-maker";
    genericName = "Procedural texture generation and 3D model painting tool";
    desktopName = "Material Maker";
    comment = finalAttrs.meta.description;
    terminal = false;
    categories = [
      "Graphics"
      "3DGraphics"
    ];
    keywords = [ "material" ];
  };

  installPhase =
    ''
      runHook preInstall
    ''
    + lib.optionalString stdenvNoCC.isLinux ''
      install -D -m 755 $src/material_maker.x86_64 $out/libexec/material-maker
      install -D -m 644 $src/material_maker.pck $out/libexec/material-maker.pck

      for dir in examples library nodes doc environments meshes export; do
        cp -R $src/$dir $out/libexec
      done

      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/

      mkdir -p $out/bin
      ln -s $out/libexec/material-maker $out/bin/material-maker
    ''
    + lib.optionalString stdenvNoCC.isDarwin ''
```

Since no C compiler is needed

eljamm commented:
I used [nurl](https://github.com/nix-community/nurl) to generate it. How did you do it?

aurreland commented:
nix-prefetch-url --unpack

aurreland commented:
nix-prefetch-url --unpack

eljamm commented:
I think yours is correct. I didn't event notice that the fetcher wasn't even supported for nurl.

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,116 @@
+    + ''
+      runHook postInstall
+    '';
+
```
```suggestion

passthru.updateScript = nix-update-script { };
```

To semi-automate update of this package. (I haven't tested if it works tho)

d-brasher reviewed (changes requested):
It could also be a good idea to split the documentation and any additional non essential file in separate outputs.

I should also note that i get the following while starting the application:

```
libudev.so.1: cannot open shared object file: No such file or directory
 
Cannot open user://export_targets
Failed to open user://library/user.json
Failed to open user://library/user_brushes.json
```

The last three lines don't seem a problem, but I don't know about the first

aurreland commented:
it throws error about not having audio and udev so i added these back as runtime dependencies

aurreland commented:
fixed this issue : libudev.so.1: cannot open shared object file: No such file or directory

aurreland commented:
We cant actually fetchzip a dmg though

aurreland commented:
If the platfrom is darwin we should fetch url then with 7z extract and move the .app folder

aurreland commented:
Should i add an unpack phase if darwin or put all in the install phase

eljamm commented:
I think it makes more sense to put it in an unpackPhase for darwin.

aurreland commented:
I didnt manage to get it working with lib.mkIf because if its not darwin unpack phase = null so is there a way to make it cleaner because a else that gives ""; is not incredible

aurreland commented:
7zip is throwing errors even tho it extracts the dmg correctly

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,138 @@
+  # build deps
+  autoPatchelfHook,
+  copyDesktopItems,
+  makeDesktopItem,
```
There is actually a setup hook for unpacking `dmg` files

```suggestion
  makeDesktopItem,
  undmg,
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,138 @@
+      ''
+    else
+      "";
+
```
```suggestion
```

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,138 @@
+    keywords = [ "material" ];
+  };
+
+  unpackPhase =
+    if stdenv.isDarwin then
+      ''
+        ${pkgs.undmg}/bin/undmg $src
+      ''
+    else
+      "";
+
```
```suggestion
```

aurreland commented:
Works well with undmg

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,138 @@
+    homepage = "https://www.materialmaker.org/";
+    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models";
+    license = lib.licenses.mit;
+    platforms = lib.platforms.darwin ++ (lib.intersectLists lib.platforms.linux lib.platforms.x86_64);
```
```suggestion
    platforms = lib.platforms.darwin ++ (lib.intersectLists lib.platforms.linux lib.platforms.x86_64);
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
```

See https://nixos.org/manual/nixpkgs/stable/#sec-meta-sourceProvenance

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,116 @@
+      runHook preInstall
+    ''
+    + lib.optionalString stdenv.isLinux ''
+      install -D -m 755 $src/material_maker.x86_64 $out/libexec/material-maker
+      install -D -m 644 $src/material_maker.pck $out/libexec/material-maker.pck
+
+      for dir in examples library nodes doc environments meshes export; do
+        cp -R $src/$dir $out/libexec
+      done
+
+      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/
+
+      mkdir -p $out/bin
+      ln -s $out/libexec/material-maker $out/bin/material-maker
```
How about `/opt` ?

d-brasher replied:
Right, that seems the right place. I forgot about it :sweat_smile:

I propose a solution below

aurreland commented:
Do you know where undmg is unpacking the dmg

eljamm commented:
you can use `sourceRoot = ".";` to tell it to unpack it in the current dirctory, I think

aurreland commented:
Works fine, it copy to the correct place. Both darwin and linux are building correctly for me.

aurreland commented:
Cant Squash the commits anymore, how can i rollback before i merged?

aurreland commented:
Hopes it works

aurreland commented:
If its Merged, Should i open a PR to backport it to latest release ?

eljamm commented:
In that case, we can just add a backport label and it will automatically take care of that. In cases where an automatic backport is not possible, you'd have to do it manually.

aurreland commented:
Is it normal the test for darwin are taking this long, are there errors in the file or is it normal behavior

eljamm commented:
I don't know why, but usually darwin builds take a longer time, so it's normal. Did it work your mac?

aurreland commented:
I didnt try on mac hardware as i dont have any available at the moment but tried switching the isDarwin to isLinux and it copied the .app to the right place so i dont see why it wouldnt work

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,133 @@
+  };
+
+  installPhase =
+    ''
+      runHook preInstall
+    ''
+    + lib.optionalString stdenv.isLinux ''
+      install -D -m 755 $src/material_maker.x86_64 $out/libexec/material-maker
+      install -D -m 644 $src/material_maker.pck $out/libexec/material-maker.pck
+
+      for dir in examples library nodes doc environments meshes export; do
+        cp -R $src/$dir $out/libexec
+      done
+
+      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/
+
+      mkdir -p $out/bin
+      ln -s $out/libexec/material-maker $out/bin/material-maker
+    ''
+    + lib.optionalString stdenv.isDarwin ''
+      mkdir -p $out/Applications
+      cp -r ./material_maker.app $out/Applications
+    ''
+    + ''
+      runHook postInstall
+    '';
```
```suggestion
    ''
      runHook preInstall
    ''
    + lib.optionalString stdenv.isLinux ''
      install -D -m 755 $src/material_maker.x86_64 $out/opt/material-maker
      install -D -m 644 $src/material_maker.pck $out/opt/material-maker.pck

      mkdir -p $out/share/
      cp -R $src/doc $out/share
    ''
    # Symlink docs if the corresponding output is installed
    # This allows opening documentation directly from the app
    + lib.optionalString (builtins.elem "doc" finalAttrs.meta.outputsToInstall or []) ''
      ln -s $doc/share/doc $out/opt/doc
    ''
    + ''
      for dir in library nodes environments meshes export; do
        cp -R $src/$dir $out/opt
      done

      cp -R $src/examples $examples

      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/

      mkdir -p $out/bin
      ln -s $out/opt/material-maker $out/bin/material-maker
    ''
    + lib.optionalString stdenv.isDarwin ''
      mkdir -p $out/Applications
      cp -r ./material_maker.app $out/Applications
    ''
    + ''
      runHook postInstall
    '';

  outputs = [ "out" "doc" "examples" ];
```

This is a proposal to use `/opt` and to modularize the package into three outputs (the main one, the documentation and the examples), in order to save disk space for a minimal install.

See https://nixos.org/manual/nixpkgs/stable/#chap-multiple-output.

I'm not sure at all that the commented passage is the correct way to have a cross-output symlink, I'd let someone more expert than me comment on this approach.

aurreland replied:
Seems definitely better than using lib exec for sure though shouldn't it all be in the lib optional isLinux bc for Darwin everything is put in the .app/ correct me if I am wrong. Maybe see  the aur for inspiration https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=material-maker-bin

aurreland replied:
Why are you querying meta.inputsToInstall but you've not added it ? and what is the point of the `or []` should add `and stdenv.isLinux` ?

d-brasher replied:
> Seems definitely better than using lib exec for sure though shouldn't it all be in the lib optional isLinux bc for Darwin everything is put in the .app/ correct me if I am wrong. Maybe see the aur for inspiration https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=material-maker-bin

Yes, you are right, I forgot `isLinux` in the second block!

> Why are you querying meta.inputsToInstall but you've not added it ? and what is the point of the `or []` should add `and stdenv.isLinux` ?

Also here, I forgot `IsLinux`. 
The `or [ ]` is needed if `outputsToInstall` is not defined (as by default it will install only the "out" output).
Probably you are right and it's better to specify `meta.outputsToInstall = [ "out" ];` and drop `or [ ]`.

(it's my first time tinkering with multiple outputs, feel free to drop the idea if you are not sure it's a good one!)

aurreland replied:
We'll maybe need to also link the examples as it is maybe used by the app, need some testing. I think it's cleaner to declare the outputsToInstall but I'm not sure. I will change some things to add the isLinux and fix the commit and will push this evening.

aurreland commented:
I added multiple outputs, thought i believe we should include everything by default, doc and examples, I added the outputsToInstall to be the outputs by default to not degrade user exprerience, correct me if im wrong. Everything works on my machine.
d-brasher could you enlighten me why when we install it we put docs in $out and not in doc but examples in $examples.
I added keyword for more clarity in the desktop item.

d-brasher commented:
> I added multiple outputs, thought i believe we should include everything by default, doc and examples, I added the outputsToInstall to be the outputs by default to not degrade user exprerience, correct me if im wrong. Everything works on my machine. d-brasher could you enlighten me why when we install it we put docs in $out and not in doc but examples in $examples. I added keyword for more clarity in the desktop item.

I don't have a strong opinion on which outputs it's better to install by default, so feel free to do as you please!

As written [here](https://nixos.org/manual/nixpkgs/stable/#multiple-output-file-type-groups) there are some paths under `$out` that are automatically recognized and moved to a separate output if specified. One of these is `share/doc/`, which goes on the `doc` output if specified.

However, `$examples` is a custom output, so we need to manually assign files to it.

(Or so that is what I understand by reading the documentation :sweat_smile:)

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,159 @@
+    + lib.optionalString (builtins.elem "doc" finalAttrs.meta.outputsToInstall && stdenv.isLinux) ''
+      ln -s $doc/share/doc $out/opt/doc
+    ''
+    +
+      lib.optionalString (builtins.elem "examples" finalAttrs.meta.outputsToInstall && stdenv.isLinux)
+        ''
+          ln -s $examples $out/opt/examples
+        ''
```
```suggestion
    + lib.optionalString (builtins.elem "examples" finalAttrs.meta.outputsToInstall && stdenv.isLinux) ''
      ln -s $examples $out/opt/examples
    ''
```

Nitpick to uniform style

d-brasher commented:
At last, you can address https://github.com/NixOS/nixpkgs/pull/331491#discussion_r1700782171 and https://github.com/NixOS/nixpkgs/pull/331491#discussion_r1702822603 (if you want) and squash the unneeded commits, then I think this is good to go!

aurreland commented:
I don't see the comment you are pointing and is your correction ok as it doesn't pass nixfmt ?

d-brasher commented:
> I don't see the comment you are pointing and is your correction ok as it doesn't pass nixfmt ?

Welp, I did not check it with `nixfmt`, sorry!

The first comment was about replacing `stdenv` with `stdenvNoCC` since we do not need the C compiler.

aurreland commented:
> The first comment was about replacing stdenv with stdenvNoCC since we do not need the C compiler.

I'll add it, I thought we already added it. 
And I think it will be ready.

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,161 @@
+    ];
+  };
+
+  unpackPhase = lib.mkIf stdenvNoCC.isDarwin ''
+    7z x $src
+  '';
```
Build fails here. Isn't this supposed to be removed?

aurreland replied:
Yes my bad seems to be a merge issue

d-brasher commented:
It now appears to miss a `nixfmt` pass (probably a trailing whitespace somewhere)

(if it can be of help, to squash commits I usually use `git rebase --interactive [latest commit on master]` and select all the commits to squash, this would prevent any merge conflict)

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,157 @@
+  };
+  
+  installPhase =
+    ''
+      runHook preInstall
+    ''
+    + lib.optionalString stdenvNoCC.isLinux ''
+      install -D -m 755 $src/material_maker.x86_64 $out/opt/material-maker
+      install -D -m 644 $src/material_maker.pck $out/opt/material-maker.pck
+
+      for dir in library nodes environments meshes export; do
+        cp -R $src/$dir $out/opt
+      done
+
+      cp -R $src/examples $examples
+
+      mkdir -p $out/share/
+      cp -R $src/doc $out/share
+
+      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/
+
+      mkdir -p $out/bin
+      ln -s $out/opt/material-maker $out/bin/material-maker
+    ''
+    + lib.optionalString (builtins.elem "doc" finalAttrs.meta.outputsToInstall && stdenvNoCC.isLinux) ''
+      ln -s $doc/share/doc $out/opt/doc
+    ''
+    +
+      lib.optionalString (builtins.elem "examples" finalAttrs.meta.outputsToInstall && stdenvNoCC.isLinux)
+        ''
+          ln -s $examples $out/opt/examples
+        ''
+    + lib.optionalString stdenvNoCC.isDarwin ''
+      mkdir -p $out/Applications
+      find ./ -type d -name "*.app" -exec cp -r {} /destination/directory $out/Applications \;
+    ''
+    + ''
+      runHook postInstall
+    '';
```
In my opinion, looping over `getAllOutputNames` is more readable.

Also, since the command that installs the icon also creates the `$outshare` directory, we can move `cp -R $src/doc $out/share` below that so we don't have to `mkdir -p $out/share/`

```suggestion
  installPhase =
    ''
      runHook preInstall
    ''
    + lib.optionalString stdenvNoCC.isLinux ''
      install -D -m 755 $src/material_maker.x86_64 $out/opt/material-maker
      install -D -m 644 $src/material_maker.pck $out/opt/material-maker.pck

      for dir in library nodes environments meshes export; do
        cp -R $src/$dir $out/opt
      done

      install -D -m 644 $src/doc/_static/icon.png -t $out/share/icons/hicolor/256x256/apps/

      mkdir -p $out/bin
      ln -s $out/opt/material-maker $out/bin/material-maker

      cp -R $src/doc $out/share
      cp -R $src/examples $examples

      for output in $(getAllOutputNames); do
        if [ "$output" == "doc" ]; then
          ln -s $doc/share/doc $out/opt/doc
        elif [ "$output" == "examples" ]; then
          ln -s $examples $out/opt/examples
        fi
      done
    ''
    + lib.optionalString stdenvNoCC.isDarwin ''
      mkdir -p $out/Applications
      find ./ -type d -name "*.app" -exec cp -r {} /destination/directory $out/Applications \;
    ''
    + ''
      runHook postInstall
    '';
```

aurreland replied:
I added it but changed it so it queries outputsToInstall instead, i think this is wat it needs to follow, correct me if im wrong

aurreland replied:
Because if the user wants to only install the out and not doc and examples they will override the meta.outputsToInstall because if you remove outputs it doesnt build anymore

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,153 @@
+      cp -r $src/doc $out/share
+      cp -r $src/examples $examples
+
+      for output in ${builtins.concatStringsSep " " finalAttrs.meta.outputsToInstall}; do
```
As I understand it, `meta.outputsToInstall` is used for the default outputs, which isn't a problem here since `doc` and `examples` are in the default already.

However, we will typically change `outputs`, so we need to match all cases from that. Also, if `meta.outputsToInstall` changes, the package will be rebuilt, and we don't want that to happen as changes to meta shouldn't cause a rebuild.

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,153 @@
+    platforms = lib.platforms.darwin ++ (lib.intersectLists lib.platforms.linux lib.platforms.x86_64);
+    maintainers = with lib.maintainers; [ aurreland ];
+    mainProgram = "material-maker";
+    outputsToInstall = finalAttrs.outputs;
```
It might be better to specify the outputs directly, here:

```suggestion
    outputsToInstall = [
      "out"
      "doc"
      "examples"
    ];
```

If `outputs` change, that shouldn't change the default outputs, right?

eljamm replied:
I was kinda think the same as well, sure.

eljamm commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,156 @@
+    icon = "material-maker";
+    genericName = "Procedural texture generation and 3D model painting tool";
+    desktopName = "Material Maker";
+    comment = finalAttrs.meta.description;
```
I think this is my final suggestion as everything else looks good.

Changes to meta shouldn't cause a rebuild. We can opt to the the comment from the upstream [desktop file](https://github.com/RodZill4/material-maker/blob/0268009d02061c566c281a0d65adde67f2f1caa1/material_maker/misc/linux/io.github.RodZill4.Material-Maker.desktop#L5) here as well

```suggestion
    comment = "Generate textures procedurally for use in PBR engines and paint 3D models";
```

eljamm reviewed (approved):
Thanks a lot @aurreland and @d-brasher for your amazing contributions.

This looks good to me 😁

d-brasher commented:
Result of `nixpkgs-review pr 331491` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>material-maker</li>
    <li>material-maker.doc</li>
    <li>material-maker.examples</li>
  </ul>
</details>

d-brasher reviewed (approved):
LGTM, too!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1881

d-brasher commented on pkgs/by-name/ma/material-maker/package.nix:
```diff
@@ -0,0 +1,156 @@
+    homepage = "https://www.materialmaker.org/";
+    description = "Tool based on Godot Engine that can be used to create textures procedurally and paint 3D models";
+    license = lib.licenses.mit;
+    platforms = lib.platforms.darwin ++ (lib.intersectLists lib.platforms.linux lib.platforms.x86_64);
```
```suggestion
    platforms = lib.platforms.darwin ++ (lib.intersectLists lib.platforms.linux lib.platforms.x86_64);
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
```

Just noticed that this got lost in the merges

aurreland commented:
I think we can remove finalAttrs, I'll do it this evening if you don't  see any problem.

aurreland commented:
Its currently not being used in the package. And I don't know if it's useful to have it in. It's difficult to find good docs about it.

aurreland commented:
And if we keep it, should we use less let in and more finalAttrs refs?

d-brasher commented:
[This issue](https://github.com/NixOS/nixpkgs/issues/315337) is the only thing I can find about `let in` vs `finalAttrs`.
It appears to be debated, but `finalAttrs` has the advantage to allow easy overrides.

You could indeed drop the use of `let in` and move everything inside `mkDerivation` using `finalAttrs` (imho it would be better), but I am by no means an expert on this.

If you choose to keep the `let in` as is, I think you can remove `finalAttrs`.

aurreland commented:
I'll look into moving everything into finalAttrs.

aurreland commented:
Does this looks good to you ? I'll squash when its finalized

d-brasher commented:
looks fine to me :+1:

sivizius commented:
Duplicate of https://github.com/NixOS/nixpkgs/pull/315483?
